File: android/src/main/java/com/scandit/datacapture/flutter/parser/ParserMethodHandler.java
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2024- Scandit AG. All rights reserved.
 */
package com.scandit.datacapture.flutter.parser;

import androidx.annotation.NonNull;

import com.scandit.datacapture.flutter.core.utils.FlutterResult;
import com.scandit.datacapture.frameworks.core.FrameworkModule;
import com.scandit.datacapture.frameworks.core.locator.ServiceLocator;
import com.scandit.datacapture.frameworks.parser.ParserModule;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class ParserMethodHandler implements MethodChannel.MethodCallHandler {

    private final ServiceLocator<FrameworkModule> serviceLocator;

    public ParserMethodHandler(ServiceLocator<FrameworkModule> serviceLocator) {
        this.serviceLocator = serviceLocator;
    }

    @Override
    public void onMethodCall(MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "parseString":
                assert call.arguments() != null;
                getSharedModule().parseString(call.arguments(), new FlutterResult(result));
                break;
            case "parseRawData":
                assert call.arguments() != null;
                getSharedModule().parseRawData(call.arguments(), new FlutterResult(result));
                break;
            case "createUpdateNativeInstance":
                assert call.arguments() != null;
                getSharedModule().createOrUpdateParser(call.arguments(), new FlutterResult(result));
                break;
            case "disposeParser":
                assert call.arguments() != null;
                getSharedModule().disposeParser(call.arguments(), new FlutterResult(result));
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private volatile ParserModule sharedModuleInstance;

    private ParserModule getSharedModule() {
        if (sharedModuleInstance == null) {
            synchronized (this) {
                if (sharedModuleInstance == null) {
                    sharedModuleInstance = (ParserModule) this.serviceLocator.resolve(ParserModule.class.getName());
                }
            }
        }
        return sharedModuleInstance;
    }
}

File: android/src/main/kotlin/com/scandit/datacapture/flutter/parser/ParserMethodHandler.kt
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2021- Scandit AG. All rights reserved.
 */

package com.scandit.datacapture.flutter.parser

import com.scandit.datacapture.flutter.core.utils.FlutterResult
import com.scandit.datacapture.frameworks.parser.ParserModule
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class ParserMethodHandler(
    private val parserModule: ParserModule
) : MethodChannel.MethodCallHandler {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            METHOD_PARSE_STRING -> parserModule.parseString(
                call.arguments as String,
                FlutterResult(result)
            )

            METHOD_PARSE_RAW_DATA -> parserModule.parseRawData(
                call.arguments as String,
                FlutterResult(result)
            )

            METHOD_CREATE_UPDATE_NATIVE_INSTANCE -> parserModule.createOrUpdateParser(
                call.arguments as String,
                FlutterResult(result)
            )

            METHOD_DISPOSE_PARSER -> parserModule.disposeParser(
                call.arguments as String,
                FlutterResult(result)
            )

            else -> result.notImplemented()
        }
    }

    companion object {
        private const val METHOD_PARSE_STRING = "parseString"
        private const val METHOD_PARSE_RAW_DATA = "parseRawData"
        private const val METHOD_CREATE_UPDATE_NATIVE_INSTANCE = "createUpdateNativeInstance"
        private const val METHOD_DISPOSE_PARSER = "disposeParser"
    }
}
